File: amy_editor/trunk/lib/amy/support/simple_http.php
<?php
/*
  *------------------------------------------------------------------------------------------
	SimpleHTTP Library

	Built on the technologies developed and maintained by April-Child.com
	Copyright (c)2007 Petr Krontorad, April-Child.com.

	Author: Petr Krontorad, petr@krontorad.com

	All rights reserved.
  *------------------------------------------------------------------------------------------

	
*/

class SimpleHTTP
{
    public static function send_post($hostname, $url_path, $data = array())
    {
        return self::send('POST', $hostname, 80, $url_path, $data);
    }
    
    public static function send_get($hostname, $url_path, $data = array())
    {
        return self::send('GET', $hostname, 80, $url_path, $data);
    }
    
	protected static function decode_body($headers, $str, $eol = "\r\n")
	{
		$tmp = $str;
		$add = strlen ($eol);
		$str = '';
		if (isset($headers['transfer-encoding']) && 'chunked' == $headers['transfer-encoding'])
		{
			do
			{
				$tmp = ltrim($tmp);
				$pos = strpos($tmp, $eol);
				$len = hexdec(substr($tmp,0,$pos));
				if (isset($headers['content-encoding']))
				{
					$str .= @gzinflate(substr($tmp, $pos + $add + 10, $len));
				}
				else
				{
					$str .= substr($tmp, $pos + $add, $len);
				}

				$tmp = substr($tmp, $len + $pos + $add);
				$check = trim ($tmp);
			}
			while (!empty($check));
		}
		else if (isset( $headers['content-encoding']))
		{
			$str = @gzinflate(substr($tmp, 10));
		}
		else 
		{
			$str = $tmp;
		}
		return $str;
	}

    public static function send($method, $hostname, $port, $url_path, $data = array(), $headers = array())
    {
        $method = strtoupper($method);
        if (!in_array($method, array('GET', 'POST', 'PROPFIND', 'OPTIONS', 'PUT', 'MKCOL')))
        {
            $method = 'GET';
        }
		if (!is_numeric($port))
		{
			$port = 80;
		}
        $s_data = array();
		if (!is_array($data))
		{
			$s_data = $data;
		}
		else
		{
	        foreach ($data as $key=>$value)
	        {
	            $s_data[] = $key.'='.urlencode($value);
	        }
	        $s_data = implode('&', $s_data);
		}
		if ('GET' == $method && 0 != strlen($s_data))
		{
			$url_path .= '?' . $s_data;
			$s_data = '';
		}
		$status_code = '';
        if (!$s = @fsockopen($hostname, $port))
		{
	        $body = '';
			$headers = array();
			$status_code = '500';
			$status = 'HTTP/1.1 500 Internal server error or host unreachable.';
		}
		else
		{
			$content_type = 'application/x-www-form-urlencoded';
			if (isset($headers['Content-Type']))
			{
				$content_type = $headers['Content-Type'];
				unset($headers['Content-Type']);
			}
	        fputs($s, $method.' '.$url_path." HTTP/1.1\r\n");
	        fputs($s, "Host: $hostname\r\n");
	        fputs($s, "Content-Type: $content_type\r\n");
	        fputs($s, "Content-Length: ".strlen($s_data)."\r\n");
	        fputs($s, "Accept-Encoding: compress, gzip\r\n");
	        fputs($s, "User-Agent: Gecko MSIE AppleWebKit KHTML Opera Win Mac Linux (April-Child.com simple HTTP client)\r\n");
			foreach ($headers as $name => $value)
			{
		        fputs($s, "$name: $value\r\n");
			}
	        fputs($s, "Connection: close\r\n\r\n");
	        fputs($s, $s_data);
	        $response = '';
	        while (!feof($s))
	        {
	            $response .= fgets($s, 128);
	        }
	        fclose($s);
	        $headers = array();
	        $body = '';
	        if (false !== $ix = strpos($response, "\r\n\r\n"))
	        {
	            $raw_headers = explode("\r\n", substr($response, 0, $ix));
	            $status = array_shift($raw_headers);
	            preg_match_all('|\\d{3}|', $status, $m);
	            $status_code = $m[0][0];
	            foreach ($raw_headers as $header)
	            {
	                $pair = explode(':', $header);
	                $headers[trim(strtolower($pair[0]))] = trim($pair[1]);
	            }
				$body = SimpleHTTP::decode_body($headers, substr($response, $ix+4));
	        }
		}
        return array('status' => $status, 'status_code' => $status_code, 'headers' => $headers, 'body' => $body);
    }    
}

File: amy_editor/trunk/api/rails_plugin/amy/lib/support/simple_http.rb
#++
#     *------------------------------------------------------------------------------------------
#     Simple HTTP Library
#
#     Built on the technologies developed and maintained by April-Child.com
#     Copyright (c)2007 Petr Krontorad, April-Child.com.
# 
#     Author: Petr Krontorad, petr@krontorad.com
# 
#     All rights reserved.
# *------------------------------------------------------------------------------------------
#--

require 'zlib'
require 'socket'
require 'cgi'
include Socket::Constants


module Amy
  module Support
    class SimpleHTTP
    
      SUPPORTED_METHODS = %w(GET POST PROPFIND OPTIONS PUT MKCOL MOVE COPY)
    
      class << self
      
        def send_post(hostname, url_path, data={})
          send('POST', hostname, 80, url_path, data)
        end
      
        def send_get(hostname, url_path, data={})
          send('GET', hostname, 80, url_path, data)
        end
      
        def send(method, hostname, port, url_path, data={}, headers={})
          method.upcase!
          method = 'GET' unless SUPPORTED_METHODS.include?(method)
          s_data = []
          unless data.respond_to?(:values)
            # plain string data sent (eg. contents of body in PUT method)
            s_data = data
          else
            data.each do |key, value|
              value = '' if value.nil?
              s_data << "#{key}=#{CGI::escape(value)}"
            end
            s_data = s_data.join('&')
          end
          if 'GET' == method && 0 != s_data.length
            url_path = url_path + "?#{s_data}"
            s_data = ''
          end
          status_code = ''
    			content_type = 'application/x-www-form-urlencoded'
    			if headers.has_key?('Content-Type')
    				content_type = headers['Content-Type']
    				headers.delete('Content-Type')
    			end

          socket = Socket.new(AF_INET, SOCK_STREAM, 0)
          begin
            sockaddr = Socket.pack_sockaddr_in(port, hostname)
            socket.connect(sockaddr)
            socket.write("#{method} #{url_path} HTTP/1.1\r\n")
            socket.write("Host: #{hostname}\r\n")
            socket.write("Content-Type: #{content_type}\r\n")
            socket.write("Content-Length: #{s_data.length}\r\n")
            socket.write("Accept-Encoding: compress, gzip\r\n")
            socket.write("User-Agent: Gecko MSIE AppleWebKit KHTML Opera Win Mac Linux (April-Child.com simple HTTP client)\r\n")
            headers.each {|name, value| socket.write("#{name}: #{value}\r\n")}
            socket.write("Connection: close\r\n\r\n")
            socket.write(s_data)
            response = socket.read
          rescue => e
            response = ''
            status = 'HTTP/1.1 500 Internal server error or host unreachable.'
            status_code = 500
          end
          headers = {}
          body = ''
          if ix = response.index("\r\n\r\n")
            raw_headers = response[0, ix].split("\r\n")
            status = raw_headers.shift
            status_code = status.match(/\d{3}/).to_s
            raw_headers.each do |header|
              pair = header.split(':')
              headers[pair[0].downcase.strip] = pair[1].strip
            end
            body = decode_body(headers, response[ix+4..-1])
          end
          {:status => status, :status_code => status_code.to_i, :headers => headers, :body => body}
        end
      
        private
          def decode_body(headers, str, eol="\r\n")
            tmp = str
            add = eol.length
            str = ''
            if headers['transfer-encoding'] && 'chunked' == headers['transfer-encoding']
              while 0 != tmp.length
                tmp = tmp.lstrip
                pos = tmp.index(eol)
                break if pos.nil?
                len = tmp[0, pos].hex
                if headers['content-encoding']
                  str << Zlib::Inflate.new(Zlib::MAX_WBITS + 32).inflate(tmp[pos+add, len])
                else
                  str << tmp[pos+add, len]
                end
                tmp = tmp[len+pos+add..-1]
              end
            elsif headers['content-encoding']
              str = Zlib::Inflate.new(Zlib::MAX_WBITS + 32).inflate(tmp[0..-1])
            else
              str = tmp
            end
            str
          end      
      end
    end
  end
end

# response = Amy::Support::SimpleHTTP.send_post('www.april-child.com', '/amy/backend/amy.php', {:a => 'list_bundles'})
# puts "Status code: #{response[:status_code]} - whole: #{response[:status]}"
# response[:headers].each {|name, value| puts "#{name}: #{value}\n"}
# puts response[:body]



# response = Amy::Support::SimpleHTTP.send_post('amy.imac', '/projects/playground.php', {:a => 'load_resource', :path => '/prototype.js'})
# puts "Status code: #{response[:status_code]} - whole: #{response[:status]}"
# response[:headers].each {|name, value| puts "#{name}: #{value}\n"}
# puts response[:body]


# response = Amy::Support::SimpleHTTP.send('POST', 'amy-rails.imac', 80, '/projects/playground', {:a => 'description'})
# puts "Status code: #{response[:status_code]} - whole: #{response[:status]}"
# response[:headers].each {|name, value| puts "#{name}: #{value}\n"}
# puts response[:body]

